include/lsst/afw/geom/Endpoint.h

```diff
@@ -0,0 +1,303 @@
+
+    /**
+    Construct a BaseEndpoint
+
+    @param[in] nAxes  the number of axes in a point
+    */
+    explicit BaseEndpoint(int nAxes) : _nAxes(nAxes) {}
+
+    virtual ~BaseEndpoint() {};
+
+    int getNAxes() const { return _nAxes; }
+
+    /**
+    Return the number of points in an array
+    */
+    virtual int getNPoints(Array const &arr) const = 0;
+
+    /**
+    Get raw data from a single point
+
+    @param[in] point  data for a single point
+    @return the values in the point as a vector of size NAxess
+
+    @throw lsst::pex::exceptions::InvalidParameterError if the point has the wrong number of axes
+    */
+    virtual std::vector<double> dataFromPoint(Point const & point) const = 0;
+
+    /**
+    Get raw data from an array of points
+    
+    @param[in] arr  Array of points
+    @return the data as a 2-D ndarray array [nPoints, nAxes] in C order,
+        so the in-memory view is, for example, x0, y0, x1, y1, x2, y2, ...
+
+    @throw lsst::pex::exceptions::InvalidParameterError if the array has the wrong nAxes dimension
+    */
+    virtual ndarray::Array<double, 2, 2> dataFromArray(Array const & arr) const = 0;
+
+    /**
+    Get a single point from raw data
+
+    @param[in] data  Data as a vector of length NAxes
+    @return the corresponding point
+    */
+    virtual Point pointFromData(std::vector<double> const & data) const = 0;
+
+    /**
+    Get an array of points from raw data
+    
+    @param[in] data  Raw data for an array of points, as a 2-D ndarray array [nPoints, nAxes] in C order,
+        so the in-memory view is, for example, x0, y0, x1, y1, x2, y2, ...
+    @return an array of points
+
+    @throw lsst::pex::exceptions::InvalidParameterError if the array has the wrong nAxes dimension
+    */
+    virtual Array arrayFromData(ndarray::Array<double, 2, 2> const & data) const = 0;
+
+    /**
+    Create a Frame that can be used with this end point in a Transform
+    */
+    virtual std::shared_ptr<ast::Frame> makeFrame() const;
+
+    /**
+    Adjust and check the frame as needed.
+
+    Do not obother to check the number of axes because that is done elsewhere.
+
+    The base implementation does nothing.
+    */
+    virtual void normalizeFrame(std::shared_ptr<ast::Frame> framePtr) const {};
+
+protected:
+    void _assertNAxes(int nAxes) const {
+        if (nAxes != this->getNAxes()) {
+            std::ostringstream os;
+            os << "number of axes provided " << nAxes << " != " << this->getNAxes() << " required";
+            throw std::invalid_argument(os.str());
+        }
+    }
+
+    int _getNAxes(ndarray::Array<double, 2, 2> const & data) const {
+        return data.getSize<1>();
+    }
+
+    int _getNAxes(ndarray::Array<double, 1, 1> const & data) const {
+        return data.getSize<0>();
+    }
+
+    int _getNAxes(std::vector<double> const & data) const {
+        return data.size();
+    }
+
+    int _getNPoints(ndarray::Array<double, 2, 2> const & data) const {
```

Comment:
No because `<double, 1, 1>` is used to represent a single point. (Also these private methods are simply used as convenience methods for other code; the fact that the code compiles and passes unit tests is a fairly strong indication that it is not missing such methods).